tests/js/client/shell/api/cursor.js

```diff
@@ -1212,9 +1212,225 @@ function dealing_with_cursorsSuite_fetching_floating_point_valuesSuite () {
+      assertEqual(doc.headers['content-type'], contentType);
+      assertTrue(doc.parsedBody.error);
+      assertEqual(doc.parsedBody.errorNum, 404);
+      assertEqual(doc.parsedBody.errorMessage, "batch id not found");
+      assertEqual(doc.parsedBody['code'], 404);
+
+      cmd = api + `/${cursorId}`;
+      doc = arango.POST_RAW(cmd, "");
+
+      assertEqual(doc.code, 200);
+      assertEqual(doc.headers['content-type'], contentType);
+      assertFalse(doc.parsedBody['error']);
+      assertEqual(doc.parsedBody['code'], 200);
+      assertFalse(doc.parsedBody['hasMore']);
+      assertEqual(doc.parsedBody['id'], undefined);
+      assertEqual(doc.parsedBody['result'].length, 1);
+      assertFalse(doc.parsedBody['cached']);
+
+      doc = arango.POST_RAW(cmd, "");
+
+      assertEqual(doc.code, internal.errors.ERROR_HTTP_NOT_FOUND.code);
+      assertEqual(doc.headers['content-type'], contentType);
+      assertTrue(doc.parsedBody['error']);
+      assertEqual(doc.parsedBody['errorNum'], internal.errors.ERROR_CURSOR_NOT_FOUND.code);
+      assertEqual(doc.parsedBody['code'], internal.errors.ERROR_HTTP_NOT_FOUND.code);
+    },
+
+    test_cursor_non_stream_retriable: function () {
+      const stmt = db._createStatement({
+        query: `FOR u IN ${cn} RETURN u`,
+        options: {stream: false, retriable: true},
+        batchSize: 100
+      });
+      let cursor = stmt.execute();
+      internal.debugSetFailAt("MakeConnectionErrorForRetry");
+
+      // the batches in between will also be retrieved with `/_api/cursor/<cursorId>/<latestBatchId>` because
+      // the failure point is located in a place in which the server would return an error, hence not returning
+      // the batch response object to the user, but it would have been constructed, so the latest batch would be cached
+      for (let i = 0; i < 2001; ++i) {
+        const nextValue = cursor.next();
+        // last batch not returning will close the cursor, won't be able to fetch the latest batch
+        if (i === 1900) {
+          internal.debugClearFailAt();
+        }
+        assertEqual("test" + i, nextValue._key);
+        assertEqual(i !== 2000, cursor.hasNext());
+      }
+      assertFalse(cursor.hasNext());
+    },
+
+    test_cursor_stream_retriable: function () {
+      const stmt = db._createStatement({
+        query: `FOR u IN ${cn} RETURN u`,
+        options: {stream: true, retriable: true},
+        batchSize: 100
+      });
+      let cursor = stmt.execute();
+      internal.debugSetFailAt("MakeConnectionErrorForRetry");
+
+      for (let i = 0; i < 2001; ++i) {
+        const nextValue = cursor.next();
+        // last batch not returning will close the cursor, won't be able to fetch the latest batch
+        if (i === 1900) {
+          internal.debugClearFailAt();
+        }
+        assertEqual("test" + i, nextValue._key);
+        assertEqual(i !== 2000, cursor.hasNext());
+      }
+      assertFalse(cursor.hasNext());
+    },
+  };
+}
```

Comment:
There should also be a test with `retriable: false`, that ensures that no `nextBatchId` is returned with the cursor results, and that a request to fetch the next id actually fails.